server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java

```diff
@@ -177,11 +183,39 @@ public LeafCollector getLeafCollector(LeafReaderContext context) throws IOExcept
                     throw e;
                 }
             }
-            aggsLeafCollector = null;
         }
-        // say that the aggs collector early terminates while the top docs collector does not, we still want to wrap in the same way
+        final LeafCollector aggsLeafCollector = alf;
+
+        if (topDocsLeafCollector == null && terminateAfter == 0 && minScore == null) {
+            // top docs collector early terminated, we can avoid wrapping as long as we don't need to apply terminate_after and min_score
+            // aggs don't support skipping low scoring hits, so we can rely on setMinCompetitiveScore being a no-op already
+            return aggsLeafCollector;
+        }
+
+        // if that the aggs collector early terminates while the top docs collector does not, we still need to wrap the leaf collector
         // to enforce that setMinCompetitiveScore is a no-op. Otherwise we may allow the top docs collector to skip non competitive
-        // hits despite the score mode of the Collector did not allow it.
+        // hits despite the score mode of the Collector did not allow it (because aggs don't support TOP_SCORES).
+        if (aggsLeafCollector == null && postFilterWeight == null && terminateAfter == 0 && minScore == null) {
```

Comment:
wow, I have read this entire function many times now. This thing is incredibly trappy. All these early exists etc. :/

Reply:
Yes, on the other hand this is all stuff that was happening before through different combinations of collector wrapper and multi collector, which was subtle. How it's all in one class, and tailored to our needs. Some weirdness surfaced too which we can address.